### PR TITLE
Push down limit through eval

### PR DESCRIPTION
**Pull request by qianheng-aws**
### Description
Add a rule EvalPushDown.PUSH_DOWN_LIMIT, this rule will push down limit under eval. Thus, limit has chance to be pushed down into TableScanBuilder later.

e.g.
```
POST _plugins/_ppl/_explain
{
  "query": """
    source=opensearch_dashboards_sample_data_flights |  eval FlightMin = FlightTimeMin | head 5 | fields  FlightMin
  """
}

# Before optimization
{
  "root": {
    "name": "ProjectOperator",
    "description": {
      "fields": "[FlightMin]"
    },
    "children": [
      {
        "name": "LimitOperator",
        "description": {
          "limit": 5,
          "offset": 0
        },
        "children": [
          {
            "name": "EvalOperator",
            "description": {
              "expressions": {
                "FlightMin": "FlightTimeMin"
              }
            },
            "children": [
              {
                "name": "OpenSearchIndexScan",
                "description": {
                  "request": """OpenSearchQueryRequest(indexName=opensearch_dashboards_sample_data_flights, 

sourceBuilder={"from":0,"size":200,"timeout":"1m"}, searchDone=false)"""
                },
                "children": []
              }
            ]
          }
        ]
      }
    ]
  }
}

# After optimization
{
  "root": {
    "name": "ProjectOperator",
    "description": {
      "fields": "[FlightMin]"
    },
    "children": [
      {
        "name": "EvalOperator",
        "description": {
          "expressions": {
            "FlightMin": "FlightTimeMin"
          }
        },
        "children": [
          {
            "name": "OpenSearchIndexScan",
            "description": {
              "request": """OpenSearchQueryRequest(indexName=opensearch_dashboards_sample_data_flights, 

sourceBuilder={"from":0,"size":5,"timeout":"1m"}, searchDone=false)"""
            },
            "children": []
          }
        ]
      }
    ]
  }
}

```

Note: Also added TODOs in this PR to implement rules PUSH_DOWN_SORT and PUSH_DOWN_PROJECT as follow-ups, which are more complex due to expression replacement.

### Related Issues
Resolves [#2903](https://github.com/opensearch-project/sql/issues/2903)

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
 - [x] New functionality has javadoc added.
 - [ ] New functionality has a user manual doc added.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.List;
+import java.util.function.BiFunction;
+import lombok.Getter;
+import lombok.experimental.Accessors;
+import org.opensearch.sql.planner.logical.LogicalEval;
+import org.opensearch.sql.planner.logical.LogicalLimit;
+import org.opensearch.sql.planner.logical.LogicalPlan;
+import org.opensearch.sql.planner.optimizer.Rule;
+
+/** */
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:26]: Javadoc missing?

**Review by dai-chen — APPROVED** [review]
Minor comment. Thanks for the changes!

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+public class EvalPushDown<T extends LogicalPlan> implements Rule<T> {
+
+  // TODO: Add more rules to push down sort and project
```
Comment (LantaoJin) [inline core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:33]: Please add some integ-tests in `ExplainIT` for current rule.
Please provide your design details for discussion in https://github.com/opensearch-project/sql/issues/2864#issuecomment-2264635556 before going to next.

Comment (qianheng-aws) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:33]: I've updated more details in the OverView of that issue https://github.com/opensearch-project/sql/issues/2864#issue-2434840370

Comment (qianheng-aws) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:33]: It seems all tests in `ExplainIT` is about sql explain. Is there any sql syntax can produce eval operator now?

Comment (LantaoJin) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:33]: > I've updated more details in the OverView of that issue [#2864 (comment)](https://github.com/opensearch-project/sql/issues/2864#issue-2434840370)

Could you create sub-issues for each task in your plan and link this PR to one of sub-issue?



> It seems all tests in `ExplainIT` is about sql explain. Is there any sql syntax can produce eval operator now?

I see. Could this rule effect the `size` parameter in `OpenSearchIndexScan`?  If yes, ExplainIT would be the proper one, or we should figure out how to verify the rule works in IT.

Comment (qianheng-aws) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:33]: > I see. Could this rule effect the size parameter in OpenSearchIndexScan? If yes, ExplainIT would be the proper one, or we should figure out how to verify the rule works in IT.

1. Sure, I'll add sub-issues for each one later.
2. Yes, it has such effect indirectly. There is another `ppl/ExplainIT` which could be the proper place to add an IT, I found the wrong one at first. I could add one. And an UT is already to verify it in `LogicalPlanOptimizerTest`.

Comment (LantaoJin) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java:33]: > I could add one. And an UT is already to verify it in LogicalPlanOptimizerTest.

I appreciate it. IT is required for any feature/enhancement, even critical bugfix.

**Review by dai-chen — APPROVED** [review]

**Comment by LantaoJin** [comment]
@qianheng-aws could you update the PR title more precise? PR title will be recorded in release notes.

**Review by LantaoJin — APPROVED** [review]
LGTM

**Pull request merged**